File: benchmark/internal/world/telecom.go
package world

import (
	"fmt"
	"strings"
)

func registerTelecom(s *Server) {
	s.handle("verify_account", func(sess *Session, args map[string]any) (any, error) {
		accounts := getList(sess.State, "accounts")
		pin := strArg(args, "pin")
		last4 := strArg(args, "last4")
		address := strArg(args, "address")
		for _, raw := range accounts {
			a, _ := raw.(map[string]any)
			if a == nil {
				continue
			}
			if asString(a["pin"]) != pin {
				continue
			}
			if last4 != "" && asString(a["last4"]) != last4 {
				continue
			}
			if address != "" && !strings.Contains(strings.ToLower(asString(a["address"])), strings.ToLower(address)) {
				continue
			}
			sess.State["identity_verified"] = true
			sess.State["account"] = a
			return map[string]any{"ok": true, "account_id": a["id"]}, nil
		}
		sess.State["identity_verified"] = false
		return map[string]any{"ok": false}, nil
	})

	s.handle("check_outage", func(sess *Session, args map[string]any) (any, error) {
		outage := truthy(sess.State["outage"])
		return map[string]any{
			"outage":  outage,
			"message": sess.State["outage_message"],
		}, nil
	})

	s.handle("walk_reboot", func(sess *Session, args map[string]any) (any, error) {
		ok := truthy(sess.State["reboot_succeeds"])
		sess.State["reboot_attempted"] = true
		sess.State["reboot_ok"] = ok
		return map[string]any{"ok": ok, "error_code": sess.State["gateway_error"]}, nil
	})

	s.handle("create_ticket", func(sess *Session, args map[string]any) (any, error) {
		if !truthy(sess.State["identity_verified"]) {
			return nil, fmt.Errorf("identity not verified")
		}
		ticket := map[string]any{
			"id":      "T-" + asString(sess.State["ticket_seq"]),
			"reason":  strArg(args, "reason"),
			"address": strArg(args, "address"),
		}
		if asString(ticket["id"]) == "T-" {
			ticket["id"] = "T-1842"
		}
		list := getList(sess.State, "tickets")
		sess.State["tickets"] = append(list, ticket)
		sess.State["ticket"] = ticket
		return ticket, nil
	})

	s.handle("dispatch_tech", func(sess *Session, args map[string]any) (any, error) {
		if !truthy(sess.State["reboot_attempted"]) {
			return nil, fmt.Errorf("reboot the gateway before dispatching a truck")
		}
		if truthy(sess.State["reboot_ok"]) {
			return nil, fmt.Errorf("service restored after reboot; no truck needed")
		}
		if !truthy(sess.State["identity_verified"]) {
			return nil, fmt.Errorf("identity not verified")
		}
		dispatch := map[string]any{
			"window": strArg(args, "window"),
			"ticket": strArg(args, "ticket_id"),
		}
		sess.State["dispatch"] = dispatch
		return dispatch, nil
	})

	s.handle("apply_credit", func(sess *Session, args map[string]any) (any, error) {
		if !truthy(sess.State["identity_verified"]) {
			return nil, fmt.Errorf("identity not verified")
		}
		if !truthy(sess.State["credit_eligible"]) {
			return nil, fmt.Errorf("account is not eligible for a bill credit")
		}
		credit := map[string]any{"amount": numArg(args, "amount")}
		sess.State["credit"] = credit
		return credit, nil
	})

	s.handle("change_plan", func(sess *Session, args map[string]any) (any, error) {
		if !truthy(sess.State["identity_verified"]) {
			return nil, fmt.Errorf("identity not verified")
		}
		plan := strArg(args, "plan")
		acct := getMap(sess.State, "account")
		acct["plan"] = plan
		sess.State["account"] = acct
		return acct, nil
	})

	s.handle("create_transfer_summary", func(sess *Session, args map[string]any) (any, error) {
		summary := strArg(args, "summary")
		sess.State["transfer_summary"] = summary
		return map[string]any{"summary": summary}, nil
	})
}

File: benchmark/agents/voicebench_agents/telecom.py
"""Telecom tools and agent factory."""

from vision_agents.core import Agent, User
from vision_agents.plugins import getstream, openai

from voicebench_agents import pack_prompt
from voicebench_agents.world_client import WorldClient


async def create_agent(**kwargs) -> Agent:
    world = WorldClient()
    llm = openai.Realtime()
    agent = Agent(
        edge=getstream.Edge(),
        agent_user=User(id="telecom-agent", name="Internet Support"),
        instructions=pack_prompt("telecom"),
        llm=llm,
    )

    @llm.register_function(
        description="Verify the account with PIN, last four, and address"
    )
    async def verify_account(pin: str, last4: str = "", address: str = "") -> dict:
        return world.call("verify_account", pin=pin, last4=last4, address=address)

    @llm.register_function(description="Check for an area outage")
    async def check_outage() -> dict:
        return world.call("check_outage")

    @llm.register_function(description="Walk the customer through a gateway reboot")
    async def walk_reboot() -> dict:
        return world.call("walk_reboot")

    @llm.register_function(description="Open a support ticket")
    async def create_ticket(reason: str, address: str = "") -> dict:
        return world.call("create_ticket", reason=reason, address=address)

    @llm.register_function(
        description="Dispatch a technician. Reboot must have failed first. window is am or pm"
    )
    async def dispatch_tech(window: str, ticket_id: str = "") -> dict:
        return world.call("dispatch_tech", window=window, ticket_id=ticket_id)

    @llm.register_function(description="Apply a bill credit if eligible")
    async def apply_credit(amount: float) -> dict:
        return world.call("apply_credit", amount=amount)

    @llm.register_function(description="Change the service plan after identity")
    async def change_plan(plan: str) -> dict:
        return world.call("change_plan", plan=plan)

    @llm.register_function(description="Store a 3-line warm transfer summary")
    async def create_transfer_summary(summary: str) -> dict:
        return world.call("create_transfer_summary", summary=summary)

    return agent
